crates/bevy_pbr/src/render/shadow_sampling.wgsl

```diff
@@ -0,0 +1,134 @@
+#define_import_path bevy_pbr::shadow_sampling
+
+#import bevy_pbr::mesh_view_bindings as view_bindings
+#import bevy_pbr::utils PI
+
+fn sample_shadow_map_hardware(light_local: vec2<f32>, depth: f32, array_index: i32) -> f32 {
+    // Do the lookup, using HW 2x2 PCF and comparison
+    // NOTE: Due to non-uniform control flow above, we must use the level variant of the texture
+    // sampler to avoid use of implicit derivatives causing possible undefined behavior.
+#ifdef NO_ARRAY_TEXTURES_SUPPORT
+    return textureSampleCompare(
+        view_bindings::directional_shadow_textures,
+        view_bindings::directional_shadow_textures_sampler,
+        light_local,
+        depth
+    );
+#else
+    return textureSampleCompareLevel(
+        view_bindings::directional_shadow_textures,
+        view_bindings::directional_shadow_textures_sampler,
+        light_local,
+        array_index,
+        depth
+    );
+#endif
+}
+
+// https://web.archive.org/web/20230210095515/http://the-witness.net/news/2013/09/shadow-mapping-summary-part-1
+fn sample_shadow_map_castano_thirteen(light_local: vec2<f32>, depth: f32, array_index: i32) -> f32 {
```

Comment:
This implements the 5x5 filter size but the implementation in MJP's has code for 3, 5, 7. We could add options for 3 and 7 later.

Reply:
5x5 I found to give good results, didn't feel the need to do 3 or 7 atm, but yeah we can add it trivially in the future (at the cost of more pipeline key bits).

Reply:
Or some view flags or something.